builtin/sparse-checkout.c

```diff
@@ -252,28 +327,53 @@ static void write_cone_to_file(FILE *fp, struct pattern_list *pl)
 
 	for (i = 0; i < sl.nr; i++) {
 		char *pattern = sl.items[i].string;
-		fprintf(fp, "/%s/\n", pattern);
+		fprintf(fp, "%s/\n", pattern);
 	}
 }
 
 static int write_patterns_and_update(struct pattern_list *pl)
 {
 	char *sparse_filename;
 	FILE *fp;
+	int fd;
+	struct lock_file lk = LOCK_INIT;
+	int result;
+	struct update_data *ud = initialize_update_data(pl);
 
 	sparse_filename = get_sparse_checkout_filename();
-	fp = fopen(sparse_filename, "w");
+	fd = hold_lock_file_for_update(&lk, sparse_filename,
+				      LOCK_DIE_ON_ERROR);
+
+	result = update_working_directory(ud);
+	if (result) {
+		rollback_lock_file(&lk);
+		free(sparse_filename);
+		clear_pattern_list(pl);
+		free(ud);
+		update_working_directory(NULL);
+		return result;
+	}
+
+	fp = fdopen(fd, "w");
 
 	if (core_sparse_checkout_cone)
 		write_cone_to_file(fp, pl);
 	else
 		write_patterns_to_file(fp, pl);
 
-	fclose(fp);
-	free(sparse_filename);
+	fflush(fp);
 
+	/* commit index.lock file */
+	finish_update_data(ud);
+	free(ud);
```

Comment:
Should the commit of the `index.lock` be the last thing to happen so after the commit of the `sparse-checkout.lock`?  I'm not sure what git reads first the sparse-checkout file or the index.  Not sure it matters too much either since both renames should happen rather fast. But since the commit has
```
  5. Commit sparse-checkout.lock.
  6. Commit index.lock.
```
Thought I would ask.

Reply:
Since the `sparse-checkout` builtin is the only thing looking at `sparse-checkout.lock`, this won't affect parallel checkouts. A `checkout`-like command would take `index.lock` before reading the `sparse-checkout` file.

To avoid deadlocks between competing `git sparse-checkout set` commands, it is best to "pop" the locks in a stack.

Reply:
And yes, I definitely did them in the wrong order in the code. Thanks for catching that!